UnityGLTF/Assets/UnityGLTF/Scripts/GLTFSceneImporter.cs

```diff
@@ -404,86 +404,90 @@ protected IEnumerator ConstructBuffer(GLTF.Schema.Buffer buffer, int bufferIndex
-                        loadedTexture.LoadImage(bufferData, true);
+                        using (MemoryStream memoryStream = stream as MemoryStream)
+                        {
 
-                        _assetCache.ImageCache[imageCacheIndex] = loadedTexture;
-                        yield return null;
+                            return memoryStream.ToArray();
+                        }
                     }
                     else
                     {
-                        Stream stream = _assetCache.ImageStreamCache[imageCacheIndex];
-                        yield return ConstructUnityTexture(stream, markGpuOnly, image, imageCacheIndex);
+                        byte[] buffer = new byte[stream.Length];
+                        // todo: potential optimization is to split stream read into multiple frames (or put it on a thread?)
+                        using (stream)
+                        {
+                            if (stream.Length > int.MaxValue)
+                            {
+                                throw new Exception("Stream is larger than can be copied into byte array");
+                            }
+                            stream.Read(buffer, 0, (int)stream.Length);
+                        }
+                        return buffer;
                     }
                 }
             }
         }
 
-        /// <summary>
-        /// Loads texture from a stream. Is responsible for stream clean up
-        /// </summary>
-        /// <param name="stream"></param>
-        /// <param name="markGpuOnly">Non-readable textures are saved only on the GPU and take up half as much memory.</param>
-        /// <returns></returns>
-        protected virtual IEnumerator ConstructUnityTexture(Stream stream, bool markGpuOnly, GLTF.Schema.Image image, int imageCacheIndex)
+        protected IEnumerator ConstructImage(GLTF.Schema.Image image, int imageCacheIndex, FilterMode filterMode, TextureWrapMode wrapMode, bool markGpuOnly = true)
         {
-            Texture2D texture = new Texture2D(0, 0);
-
-            if (stream is MemoryStream)
+            if (_assetCache.ImageCache[imageCacheIndex] == null)
             {
-                using (MemoryStream memoryStream = stream as MemoryStream)
-                {
-                    //	NOTE: the second parameter of LoadImage() marks non-readable, but we can't mark it until after we call Apply()
-                    texture.LoadImage(memoryStream.ToArray(), false);
-                }
-
+                byte[] data = GetImageDataBuffer(image, imageCacheIndex);
                 yield return null;
-            }
-            else
-            {
-                byte[] buffer = new byte[stream.Length];
 
-                // todo: potential optimization is to split stream read into multiple frames (or put it on a thread?)
-                using (stream)
+                Texture2D texture;
+                var info = DDSHeader.Read(data);

```

Comment:
We want to have a more robust extension system in the future that would make adding these easier to do.

For now though I think we would want different functions called based on the extensions on the textures